### ISSUE TITLE
Generating wrapper functions directly from Rust is the wrong approach after all

### ISSUE THREAD

**Issue by clauswilke** [extendr/extendr#123]
We need OS and build-specific modifications to the wrapper functions, and making this modifications from Rust at compile time doesn't make sense. See: https://github.com/extendr/rextendr/issues/21#issuecomment-759415572

Instead of generating wrapper functions, we should just generate a csv file or similar that contains the info required to put together the wrapper functions.

Relevant extendr-macros code is here: https://github.com/extendr/extendr/blob/master/extendr-macros/src/output_r.rs

**Comment by clauswilke** [comment]
cc @hobofan

**Comment by Ilia-Kosenkov** [comment]
Ideally (not sure if possible) it could be nice to have a feature-guard.
In `rextendr` the `Cargo.toml` file is written on the fly, so any features can be enabled for `extendr-macros`.

In a normal scenario `extendr-macros` will work as it does now. With the feature enabled, instead of writing `extendr_wrapper.R`, we output some `.csv` database that we can later consume.
This will not change public API (i.e. how macros are used in rust code), but will allow us to tweak the output and make better derived products.

**Comment by clauswilke** [comment]
There's really no reason not to just write both all the time.

**Comment by CGMossa** [comment]
I still believe that generating R wrapper functions from Rust is the right approach. I've been tinkering with it,
and found that a few modifcations to `extendr-macros` enables adding the `PACKAGE` document.

I'm writing the steps here, but will make a PR once I get os-detection correct. 

## Steps

In `write_r_wrapper` add

```rust
    let package_name = std::env::var("CARGO_CRATE_NAME")?;
```

In `to_r_wrapper` write:

```rust
    fn to_r_wrapper(&self, package_name: String) -> String { // change signature
    
    
            ".Call(\"{name}\"{leading_comma}{args}, PACKAGE = \"{package_name}\")", // amend the formatting
            
            
            package_name = package_name, // at the end of the format!-call

**Comment by andy-thomason** [comment]
I will eventually get off my backside and write the metadata generator.

This will give you a list of impls and functions that you need to generate the .Call.

Note that with the new registration, you can't use "warp__rust_inc" but must use the symbol.

**Comment by clauswilke** [comment]
@CGMossa @andy-thomason You're both missing a few things.

1. There are two separate use cases, (i) writing an R package containing Rust code and (ii) using `rust_source()` from rextendr. They need different wrapper functions.

2. The cargo crate name is not the correct package name when using `rust_source()`. That's what @Ilia-Kosenkov discovered and recently addressed in the rextendr code.

3. You don't need symbols in the wrapper functions in either case. For `rust_source()`, we're skipping routine registration. For packages, compilation turns the strings into symbols. You can see this from the fact that this wrapper works in the extendrtests package: https://github.com/extendr/extendr/blob/master/tests/extendrtests/R/wrappers.R even though directly calling the wrapper functions requires us to use symbols: https://github.com/extendr/extendr/blob/f714309fc6806f953cf82a6a30b0863a5d4eff1d/tests/extendrtests/tests/testthat/test-type-conversion.R#L2

**Comment by andy-thomason** [comment]
I'm doing a first pass at some metadata.

There will be a get_xxx_metadata function that gives the Rust metadata
and a wrapper wrap__get_xxx_metadata which returns a list.

```
/// Metadata function argument.
pub struct Arg {
    pub doc: &'static str,
    pub name: &'static str,
    pub arg_type: &'static str,
}

/// Metadata function.
pub struct Func {
    pub doc: &'static str,
    pub name: &'static str,
    pub args: Vec<Arg>,
    pub return_type: &'static str,
    pub func_ptr: *const u8,
}

/// Metadata struct.
pub struct Struct {
    pub doc: &'static str,
    pub name: &'static str,
    pub methods: Vec<Func>,
}

/// Module metadata.
pub struct Metadata {
    pub doc: &'static str,
    pub functions: Vec<Func>,
    pub structs: Vec<Struct>,
}
```

**Comment by andy-thomason** [comment]
We should be able to drop the `init__` functions once this is done.

You should then be able to write R code to generate wrappers.

**Comment by andy-thomason** [comment]
Is everyone happy with this, now?

Can we close it?

**Comment by clauswilke** [comment]
Yes, I think the current framework is fine. I still need to finish porting rextendr, but there is no fundamental technical issue.
